src/wallet/wallet.h

```diff
@@ -741,6 +745,106 @@ class CAccountingEntry
 };
 
+enum NoteType {
+    sapling,
+    sprout
+};
+
+class NotesHelperObject
+{
+public:
+
+    NotesHelperObject(uint256 hash_, uint256 address_, uint64_t timestamp_, NoteType type_,
+        boost::optional<JSOutPoint> jsop_, boost::optional<SaplingOutPoint> op_) :
+    hash(hash_), address(address_), timestamp(timestamp_), type(type_), jsop(jsop_), op(op_)
+    {}
+
+    uint256 hash;
+    uint256 address;
+    uint64_t timestamp;
+    NoteType type;
+    boost::optional<JSOutPoint> jsop;
+    boost::optional<SaplingOutPoint> op;
+};
+
+using namespace boost::multi_index;
+
+struct by_hash;
+struct by_timestamp;
+
+class NotesHelperIndex
+{
+public:
+    typedef boost::multi_index_container<
+        NotesHelperObject,
+        indexed_by<
+            ordered_non_unique<tag<by_hash>,
+                composite_key<
+                    NotesHelperObject,
+                    member<NotesHelperObject, NoteType, &NotesHelperObject::type>,
+                    member<NotesHelperObject, uint256, &NotesHelperObject::address>,
+                    member<NotesHelperObject, uint256, &NotesHelperObject::hash>
+                >
+            >,
+            ordered_unique<tag<by_timestamp>,
+                composite_key<
+                    NotesHelperObject,
+                    member<NotesHelperObject, NoteType, &NotesHelperObject::type>,
+                    member<NotesHelperObject, uint256, &NotesHelperObject::address>,
+                    member<NotesHelperObject, uint64_t, &NotesHelperObject::timestamp>
+                >
+            >
+        >
+    > notes_index;
+
+    notes_index index;
+    typedef notes_index::nth_index<1>::type::iterator by_timestamp_itr;
+
+    void insert(uint256 hash, uint256 address, uint64_t time, NoteType type, boost::optional<JSOutPoint> jsop,
+        boost::optional<SaplingOutPoint> op)
+    {
+        auto& hashaddreess_index = index.get<by_hash>();
+        auto search = hashaddreess_index.find(make_tuple(type, address, hash));
+        if (search != hashaddreess_index.end()) {
+            hashaddreess_index.replace(search, NotesHelperObject(hash, address, time, type, jsop, op));
+        }
+        else {
+            hashaddreess_index.insert(NotesHelperObject(hash, address, time, type, jsop, op));
+        }
+    }
+
+    static NotesHelperIndex* getInstance() {
+        if (instance == NULL)
+            instance = new NotesHelperIndex();
+        return instance;
+    }
+
+private:
+    NotesHelperIndex() {};
+    static NotesHelperIndex* instance;
+};
+
+class NotesFilters
+{
+public:
+    int minDepth = 1;
+    int maxDepth = INT_MAX;
+    CWalletTx* wtx;
+    CWallet* cwallet;
+    boost::optional<SproutNoteData> nd_sprout;
+    boost::optional<libzcash::SproutPaymentAddress> pa_sprout;
+    boost::optional<JSOutPoint> jsop;
+    boost::optional<SaplingNoteData> nd_sapling;
+    boost::optional<libzcash::SaplingPaymentAddress> pa_sapling;
+    boost::optional<SaplingOutPoint> op;
```

Comment:
The suggested refactoring would junk all the optional fields, retaining only the fields that determine what should pass the filter.

Reply:
https://github.com/zcash/zcash/pull/4399/commits/1431921924f32190b4f34f7c8cf1ea81d2957549